### PR TITLE
citra_qt: Add Set Play Coins option

### LINKED ISSUE #1989: Add ability to set play coins either through emulated walking or an option in the configuration window

### PR DESCRIPTION
**Pull request by valentinvanelslande**
Screenshot of the dialog:
dialog

Closes #1989

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3653)

### CODE REVIEW

File: src/core/hle/service/ptm/ptm.cpp [hunk 1]
```diff
@@ -169,6 +177,43 @@ Module::Module() {
...
+
+        // Open it again to get a valid archive now that the folder exists
+        archive_result =
+            Service::FS::OpenArchive(Service::FS::ArchiveIdCode::SharedExtSaveData, archive_path);
+    }
+
+    ASSERT_MSG(archive_result.Succeeded(), "Could not open the PTM SharedExtSaveData archive!");
+
+    FileSys::Path gamecoin_path("/gamecoin.dat");
+    Service::FS::CreateFileInArchive(*archive_result, gamecoin_path, sizeof(GameCoin));
```
Comment (wwylele) [inline src/core/hle/service/ptm/ptm.cpp]: I think both archive formatting and file creation can be omitted as it is already done in the `Module` constructor.

File: src/core/hle/service/ptm/ptm.cpp [hunk 2]
```diff
@@ -24,6 +24,14 @@ static const GameCoin default_game_coin = {0x4F00, 42, 0, 0, 0, 2014, 12, 29};
 /// Id of the SharedExtData archive used by the PTM process
 static const std::vector<u8> ptm_shared_extdata_id = {0, 0, 0, 0, 0x0B, 0, 0, 0xF0, 0, 0, 0, 0};
 
+static std::weak_ptr<Module> current_ptm;
```
Comment (lioncash) [inline src/core/hle/service/ptm/ptm.cpp]: Please don't introduce non-const global state like this. We're trying to eliminate that. Use a proper interface that uses instance state.

File: src/citra_qt/main.cpp [hunk 3]
```diff
@@ -1037,6 +1042,15 @@ void GMainWindow::OnConfigure() {
     }
 }
 
+void GMainWindow::OnSetPlayCoins() {
+    bool ok;
+    u16 play_coins = static_cast<u16>(
+        QInputDialog::getInt(this, tr("Set Play Coins"), "Play Coins:", 0, 0, 300, 1, &ok,
```
Comment (lioncash) [inline src/citra_qt/main.cpp]: Both messages should be translatable.

**Comment by wwylele** [comment]
What's the status of this PR?

File: src/core/hle/service/ptm/ptm.cpp (hunk 1 above)
Comment (valentinvanelslande) [reply src/core/hle/service/ptm/ptm.cpp]: it is possible that the file does not exist when the user sets the Play Coins

File: src/citra_qt/main.ui [hunk 4]
```diff
@@ -45,7 +45,7 @@
      <x>0</x>
      <y>0</y>
      <width>1081</width>
-     <height>26</height>
+     <height>21</height>
```
[the remark this answers is not in the document]
Comment (valentinvanelslande) [reply src/citra_qt/main.ui]: Qt Creator changed it

File: src/core/hle/service/ptm/ptm.cpp (hunk 1 above)
Comment (wwylele) [reply src/core/hle/service/ptm/ptm.cpp]: @valentinvanelslande but, again, it is already done in the constructor.

**Comment by purpasmart96** [comment]
Maybe we could just have a Misc tab  for things too small to be their own tab...

**Comment by jroweboy** [comment]
First off, I apologize for taking so long to getting around to review this!

That said, I am not sold on the design of this as it currently stands. A popup dialog is pretty unnecessary in my opinion. We could just as easily have an option in the System Settings dialog for the default number of play coins. Since settings are applied on application launch, this should be sufficient to reset the play coins to whatever amount the user chooses each time. Then in the same code path where we update the rest of the system settings, we can do the same for play coins. This would remove the need to have a current PTM module already created, and would fit in well with the rest of the UI.

for reference, we can probably make a ptm module here like we did with cfg here https://github.com/citra-emu/citra/blob/master/src/citra_qt/configuration/configure_system.cpp#L52 and use the same set play coins method that you already wrote.

If you plan on reworking this, please just reopen this PR and pick it back up :) I think the ability to set play coins is a good idea, and would love to work with you to get this merged. In the meantime, I'll be closing this as the PR queue is far too packed right now. Thanks!

**Pull request closed without merging**